plugins/catalog-backend/migrations/20201123205611_relations_table_uniq.js

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright 2020 Spotify AB
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// @ts-check
+
+/**
+ * @param {import('knex')} knex
+ */
+exports.up = async function up(knex) {
+  if (knex.client.config.client === 'sqlite3') {
+    // sqlite doesn't support dropPrimary so we recreate it properly instead
+    await knex.schema.dropTable('entities_relations');
+    await knex.schema.createTable('entities_relations', table => {
+      table.comment('All relations between entities in the catalog');
+      table
+        .uuid('originating_entity_id')
+        .references('id')
+        .inTable('entities')
+        .onDelete('CASCADE')
+        .notNullable()
+        .comment('The entity that provided the relation');
+      table
+        .string('source_full_name')
+        .notNullable()
+        .comment('The full name of the source entity of the relation');
+      table
+        .string('type')
+        .notNullable()
+        .comment('The type of the relation between the entities');
+      table
+        .string('target_full_name')
+        .notNullable()
+        .comment('The full name of the target entity of the relation');
+      table.index('originating_entity_id', 'originating_entity_id_idx');
+      table.index('source_full_name', 'source_full_name_idx');
+    });
+  } else {
+    await knex.schema.alterTable('entities_relations', table => {
+      table.dropPrimary();
+      table.index('originating_entity_id', 'originating_entity_id_idx');
+      table.index('source_full_name', 'source_full_name_idx');
```

Comment:
What's the purpose of this change? Why have and index for the originating entity ID, and why remove the indexing of the type and target full name?

Reply:
Ah the originating entity ID is a mistake - it should already be indexed by virtue of being an FK constraint. Fixing.

As for the other one, the source name is the [primary means of joining / lookup](https://github.com/backstage/backstage/blob/master/plugins/catalog-backend/src/database/CommonDatabase.ts#L504). The others aren't used in a fashion that requires an index, and I just didn't want to have too many indices for insert-performance reasons. For lookups, the current system at least is to use the `entities_search` table to find matches, not other indexed lookups.

Removing the compound constraint is due to local processing constraints - a group A that says it has a child B, while B says it has parent A, will emit duplicate parent/child relations.